Clean up a number of small issues and improve error checking

afontenot opened the pull request:
I'm piling a large number of commits here that basically just improve code quality, with no (intended) changes in behavior (other than fixing several probable bugs caused by typos).

Summary:

 * When methods of `urllib.parse` are used, it's incorrect to only import `urllib` and in fact this only worked because importing the `request` module fixed it
 * Fixed several easily overlooked issues, e.g. typo `==` for `=`, missing reference to `self` on a method, accidentally using an undefined setting key with no default value (which happened to work because `None` is falsey, and False was the default behavior). Removed some code that looked broken and dead as well.
 * Clean up cases where `.get()` was used as a non-failing way to get the value for some key on an object, but then the result immediately fails with a type error if `.get()` returns `None`. These errors are harder to understand and debug than the clear `KeyError: 'key'` or equivalent.
 * Clean up a ton of cases where a subclass implementation had a different API for a method than its parent, causing breakage if something called a method on an instance of the class while assuming the parent API.
 * Try to improve the error handling by dealing with all the cases where type annotations in dependencies indicated that we do not deal with all error conditions. `BeautifulSoup` is very bad about silently running `None` and depending on the situation an error like this may not be caught for many lines, making debugging more difficult.

The last of these changes is the most involved, and it's possible I went a little too far to make the type checker happy, e.g. dealing with cases that are impossible (I think!) like `soup.find("TagName")` returning a `NavigableString`. There are also a few cases where existing error checking (with `AttributeError`) which *did* look sufficiently robust was replaced with a solution that the type checker could analyze instead. If need be some of these changes could be reverted.

afontenot commented:
I realize reviewing this is a big ask. No hurry of course. Also, (a) I'm happy to help with anything I can in getting this in a mergeable state, and (b) I'm happy to see selected commits from this branch cherry-picked into main, and I can rebase my changes and force push here once that happens.

thisisparker commented:
This is amazing and greatly appreciated. I do want to review it with some care but I anticipate I will probably merge all these commits, maybe with some very minor changes.

(It probably goes without saying, but this project started with pretty humble ambitions and my skills have grown along with it! With the exception of a pretty major refactor a few years back, it hasn't gotten this kind of holistic review, and I am very grateful for it coming from a pair of "outside" eyes.)

afontenot commented:
Great! Was also meaning to ask whether there is a particular code style you prefer to use for this repository? There seems to be a mix of styles from what I've seen, but I'd like to clean up at least the code I've touched for consistency. If you don't have a preference my go-to is `ruff format` with [ruff](https://github.com/astral-sh/ruff).

afontenot commented:
I've added a few more commits. Other than the usual cleaning up of nits I've noticed, there's the following two significant changes:

 * I refactored the behavior of features like authentication and embedded URLs so that `xword_dl.py` is entirely neutral with respect to downloader plugins. Any plugin that offers the requisite features will be allowed to use the authentication option, for example.
 * Building on this, I refactored the way importing plugins worked pretty significantly, in a way that gets rid of a lot of the introspection required previously. Now the downloader package only makes public the function `get_plugins()`, and (for now) `xword_dl.py` uses this to populate a list of the available plugin classes. With this change the downloader package's `__init__.py` is *much* cleaner (IMHO), and the none of the code needs the `inspect` module any more. (As a result of this and some of the small changes, Pyright issues *zero* complaints for xword-dl now.)

I think it would be quite nice to try to refactor some of the download logic out to a separate file, if you're okay with that idea. My idea for how to structure this is to have only `main()` in `xword_dl.py` (along with helper functions), and have `main()` instantiate a `Downloader` class. With this class, we can make the importing of plugins more explicit (instead of a file global), and centralize functions like filtering downloader plugins that are currently spread throughout the file.

Other than a few small issues (e.g. the single-file utils package with a compressed namespace), the project is linter clean as well! The only other significant change I wanted was a little more error checking around `requests`.

afontenot commented:
Note to self: need to fixup all the 3.10+ style typing annotations `type | type` to use `Union` instead before this gets merged.

afontenot commented:
Just FYI, I'm pushing pause on this until you have time to review the changes (no pressure). There's already way more than enough here. I think the headline feature is that the codebase now passes pylint and pyright other than a couple of nits I was unable to fix without heavy refactoring.

I still plan to make the type annotations Python 3.9 compatible before this gets merged, but that fixup can happen after review.

The one significant future change I think would be helpful to add to this would be factoring the puzzle handling code out of xword_dl.py and into its own file. I'd like to make it possible to plug in alternative file formats, like ipuz (libipuz), by implementing the same puzzle API as puzpy.

afontenot commented:
> I still plan to make the type annotations Python 3.9 compatible before this gets merged

This might no longer be needed, Python 3.9 will be EOL in October.

thisisparker commented:
I would like to merge this and I'm very sorry it took so long! (I'm glad you saw that that I'm back in it, since I've been at Recurse Center for the past few months.) The only merge conflicts are around the Puzzmo downloader, which I think I can resolve by just linting my newer version.

It feels funny to ask this now, but do you have any issues with my merging this as-is? Obviously I wouldn't have needed to ask last year but it's been so long I feel like I want to give you a chance to chime in!

thisisparker commented:
Actually I have two questions, if you don't mind!
- the `__get_subclasses` function in `__init__.py` has `[T]` which I think is a [type parameter list](https://docs.python.org/3/reference/compound_stmts.html#type-params), new in 3.12? Sorry, I'll admit I'm not that fluent with the type hinting stuff. As it stands, the automated status checker is choking on it using 3.11. What do you think would be the best way to resolve that?
- again, have to plead some ignorance here, but if you'd be willing to explain the move to `classmethod` for existing `staticmethod`s I would love to hear!

thisisparker commented:
I spotted one teeny thing with the keyword listing, and it looks like the CI is stumbling on something having to do with the `matches_url` changes, on some but not all outlets. I'm going to take a look, but do you know why that might be?

thisisparker commented:
More context as I'm figuring it out: I think the issue is that the `supported_downloaders` comprehension wants to filter out downloaders that have not implemented their own `matches_url` function. Instead, it's matching everything because everything is inheriting the `matches_url` from BaseDownloader. I think I have a fix but I want to make sure it passes muster...

thisisparker commented:
Alright, how'd I do? CI won't mark as "passing" because it doesn't have access to my secrets in your fork, but I can independently confirm it works on NYT puzzles.

I did this in the way I thought was neatest, which was adding another filter to the `get_supported_outlets` function. It is a little finicky to make sure there's a subclass implementation of the function but I think this works alright! I also ran it through ruff format, which I agree, would be nice to implement across the repo.
